cmus.c

```diff
@@ -388,3 +376,72 @@ int cmus_playlist_for_each(const char *buf, int size, int reverse,
 	}
 	return 0;
 }
+
+/* multit-hreaded next track requests */
+
+int cmus_next_track_request_fd;
+static int cmus_next_track_request_fd_priv;
+static pthread_mutex_t cmus_next_file_mutex = CMUS_MUTEX_INITIALIZER;
+static pthread_cond_t cmus_next_file_cond = CMUS_COND_INITIALIZER;
+static int cmus_next_file_provided;
+static struct track_info *cmus_next_file;
+
+#define cmus_next_file_lock() cmus_mutex_lock(&cmus_next_file_mutex)
+#define cmus_next_file_unlock() cmus_mutex_unlock(&cmus_next_file_mutex)
+
+static struct track_info *cmus_get_next_from_main_thread(void)
+{
+	struct track_info *ti = play_queue_remove();
+	if (!ti)
+		ti = play_library ? lib_goto_next() : pl_goto_next();
+	return ti;
+}
+
+static struct track_info *cmus_get_next_from_other_thread(void)
+{
+	static pthread_mutex_t mutex = CMUS_MUTEX_INITIALIZER;
+
+	cmus_mutex_lock(&mutex);
+
+	char buf = 0;
+	write(cmus_next_track_request_fd_priv, &buf, 1);
```

Comment:
I'd prefer to have some `notify`-style wrappers around pipe-related read/write calls – it would make the code much more readable.